x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/checkpoint/DataFrameTransformsCheckpointService.java

```diff
@@ -0,0 +1,124 @@
+
+    /**
+     * Get a checkpoint with an id, used to store a checkpoint.
+     *
+     * @param transformConfig the @link{DataFrameTransformConfig}
+     * @param listener listener to call after inner request returned
+     */
+    public void getCheckpoint(DataFrameTransformConfig transformConfig, long checkpointId,
+            ActionListener<DataFrameTransformCheckpoint> listener) {
+        long timestamp = System.currentTimeMillis();
+
+        // placeholder for time based synchronization
+        long timeUpperBound = 0;
+
+        // 1st get index to see the indexes the user has access to
+        GetIndexRequest getIndexRequest = new GetIndexRequest().indices(transformConfig.getSource());
+
+        ClientHelper.executeWithHeadersAsync(transformConfig.getHeaders(), ClientHelper.DATA_FRAME_ORIGIN, client, GetIndexAction.INSTANCE,
+                getIndexRequest, ActionListener.wrap(getIndexResponse -> {
+                    Set<String> userIndices = new HashSet<>(Arrays.asList(getIndexResponse.getIndices()));
+
+                    // 2nd get stats request
+                    ClientHelper.executeAsyncWithOrigin(client, ClientHelper.DATA_FRAME_ORIGIN, IndicesStatsAction.INSTANCE,
+                            new IndicesStatsRequest().indices(transformConfig.getSource()), ActionListener.wrap(response -> {
+                                if (response.getFailedShards() != 0) {
+                                    throw new CheckpointException("Source has [" + response.getFailedShards() + "] failed shards");
+                                }
+
+                                Map<String, long[]> checkpointsByIndex = extractIndexCheckPoints(response.getShards(), userIndices);
+                                DataFrameTransformCheckpoint checkpointDoc = new DataFrameTransformCheckpoint(transformConfig.getId(),
+                                        timestamp, checkpointId, checkpointsByIndex, timeUpperBound);
+                                listener.onResponse(checkpointDoc);
+
+                            }, IndicesStatsRequestException -> {
+                                throw new CheckpointException("Failed to retrieve indices stats", IndicesStatsRequestException);
+                            }));
+
+                }, getIndexException -> {
+                    throw new CheckpointException("Failed to retrieve list of indices", getIndexException);
+                }));
+
+    }
+
+    static Map<String, long[]> extractIndexCheckPoints(ShardStats[] shards, Set<String> userIndices) {
+        Map<String, long[]> checkpointsByIndex = new TreeMap<>();
+
+        // go through the list of indices and extract the corresponding checkpoints
+        // beware: this needs #indices * #shards runs
+        for (String indexName : userIndices) {
+            TreeMap<Integer, Long> checkpoints = new TreeMap<>();
+            for (ShardStats shard : shards) {
+                if (shard.getShardRouting().getIndexName().equals(indexName)) {
```

Comment:
I am collecting all checkpoints of 1 index at a time, see line 118, so it's not that easy unfortunately.

However I agree this could be re-written in a way that it requires only 1 pass. I need to use better helper structures for that. I will take a look again.

Reply:
Re-ordering the check will improve readability as well. Then you can just reference `shard.getShardRouting().getIndexName()` in the code lower down instead of `indexName`

Reply:
I re-wrote this to not require 2 passes